### ISSUE TITLE
System themes

### ISSUE THREAD

**Issue by mmstick** [hecrj/iced#1022]
What's the stance on supporting system themes? It's often important to have a consistent UX across the desktop, with all applications sharing a common system theme by default. Will iced support allowing users and distributors to design system themes that can be applied universally to all iced applications?

**Comment by yusdacra** [comment]
I'm not sure how feasible it would be, but perhaps writing iced backends for native toolkits of a platform (qt, gtk, etc.) would work? I know druid supports native platform toolkits so it should be possible.

**Comment by mmstick** [comment]
Let's say that I have an OS, and I want to build that OS around a toolkit that isn't GTK or Qt. I'm already using GTK for building applications in this OS, but I want to see if there are better solutions that I could use should I want to transition away from GTK. So what I want isn't to use a toolkit that has a reliance on GTK or Qt, but a toolkit that I can use as an alternative to them. But that would require that such theoretical toolkit supports defining system themes.

**Comment by yusdacra** [comment]
Ah so you mean as in, say, something like a "theming backend", which could perhaps support using GTK themes, or Qt themes etc.?

**Comment by mmstick** [comment]
Having something like `/etc/iced/settings.{ron|toml}` that supports defining a theme by name, and having themes placed in `/usr/share/themes/{NameOfTheme}/iced/`, using whatever format works best for iced. This way, if I develop an application for this OS, it would use the system theme by default, but users could change the system theme to one they created. Applications could extend the theme or opt into their own custom theme at the application level.

**Comment by mmstick** [comment]
I guess to be clear, support for GTK and Qt themes is not something I'd expect. Instead, a theme format specific to Iced. I don't generally think it wise to try to guess color schemes and layouts from a GTK theme.

**Comment by hecrj** [comment]
I think it sounds like a good idea! I am not happy with our current default theme. This would be part of the efforts to improve it.

We could make `iced` look into user settings to obtain the default theme, instead of using the currently hardcoded one. For this, we would need to change the widgets slightly and pass the default theme around when rendering, since it will be chosen at runtime now.

**Comment by 13r0ck** [comment]
Where are the hard coded values? I am willing to try and make a draft for this functionality.

**Comment by emann** [comment]
As mentioned in https://github.com/iced-rs/iced/discussions/1257, I'd love to take part in implementing this by tackling an important precursor, which is passing around a default styling/theme at runtime. The right place seems to be adding a `widget_stylesheet: dyn WidgetStyleSheet` to `renderer::Style` for each widget, but I'm not sure. Doing some experimenting today but would love some input/ideas

**Comment by emann** [comment]
## Demo Time!

### Explanation
Took the `styling` example project and created an [`application_theme` example](https://github.com/emann/iced/tree/feature/styling-passed-around-at-runtime/examples/application_theme). Unlike the `styling` example, the widgets don't require individual calls to `.style(SomeStyleSheet)` other than the one button that does so in order to show that custom styling still works. All of the structs that implement the various widgets' respective `StyleSheet` traits are still present but they are used to create a `renderer::Style` that is returned by `styling()` and passed through to all `Widget::draw()` calls. The `.styling()` impl for the root struct that implements `Sandbox` is simply:
```rust
fn styling(&self) -> Style { self.theme.into() }
```
as the `Theme` enum has an `impl From<Theme> for Style` which returns the `renderer::Style` for the theme selected.

---

### ToDo

- Rename `renderer::Style` to `renderer::Theme` and update params/docs/etc. accordingly
- Make sure all examples still work (or at least the ones that work on master)
- Maybe remove the `styling` example as it doesn't really show anything special anymore

**If/when this gets merged in all that will be left is to create platform-specific `impl StyleSheet for Default` for the various widgets.**

---

### Be warned - the ugly theme really is **_ugly_**

![application_theme_demo](https://user-images.githubusercontent.com/16512425/156871334-344f7cc6-4bb4-475b-9933-fac081fc1df8.gif)

**Comment by oknozor** [comment]
Hey what's the current status for this @emann ? I'd could really use this. If I may suggest it would be really nice to have just a tiny subset of css (using pest or nom) to configure the themes. I am currently using toml/ron to let user configure iced styling for an app. Not having color syntax highlight is a pain.

**Comment by 13r0ck** [comment]
@oknozor https://github.com/iced-rs/rfcs/pull/6

**Comment by Frostie314159** [comment]
Just to add to this. There is a crate called [dark-light](https://crates.io/crates/dark-light) which can detect the system theme across all platforms(not sure about web though). Using it might be feasible.

**Comment by lucatrv** [comment]
The following sample code, using the [dark-light](https://crates.io/crates/dark-light) crate, works but detects system theme only once when the app is launched. It would be nice to be able to detect system theme changes at runtime, to always keep the app theme aligned to it.

```rust
use iced::{Element, Sandbox, Settings, Theme};

pub fn main() -> iced::Result {
    Hello::run(Settings::default())
}

struct Hello;

impl Sandbox for Hello {
    type Message = ();

    fn new() -> Self {
        Self
    }

    fn title(&self) -> String {
        String::from("A cool application")
    }

    fn update(&mut self, _message: Self::Message) {}

    fn view(&self) -> Element<Self::Message> {
        "Hello, world!".into()
    }

    fn theme(&self) -> Theme {
        match dark_light::detect() {
            dark_light::Mode::Light | dark_light::Mode::Default => Theme::Light,
            dark_light::Mode::Dark => Theme::Dark,
        }
    }
}
```

**Comment by ModProg** [comment]
> It would be nice to be able to detect system theme changes automatically and keep the app theme aligned to it.

I implemented `Application::subscription`:
```rs
fn subscription(&self) -> iced::Subscription<Self::Message> {
    iced_native::subscription::events().map(Message::EventOccurred)
}
```
And that meant it would auto update. Though I'm not sure what the implications of that are.

**Comment by hecrj** [comment]
It should be fairly easy to create a [`subscription::channel`](https://docs.rs/iced/latest/iced/subscription/fn.channel.html) that runs `dark_light::detect` in a `loop` every minute or so and publishes `Mode` changes to the application.

**Comment by lucatrv** [comment]
I think I finally found my solution, I report it here in case it was useful for other ones:

```rust
use iced::widget::text;
use iced::{event, executor, subscription, time, window};
use iced::{Application, Command, Element, Event, Settings, Subscription, Theme};
use std::path::PathBuf;

pub fn main() -> iced::Result {
    FileDrop::run(Settings::default())
}

struct FileDrop {
    text: String,
    theme: Theme,
}

#[derive(Debug)]
enum Message {
    FileDropped(PathBuf),
    SystemThemeMode(Theme),
}

impl Application for FileDrop {
    type Executor = executor::Default;
    type Message = Message;
    type Theme = Theme;
    type Flags = ();

    fn new(_flags: Self::Flags) -> (FileDrop, Command<Self::Message>) {
        (
            FileDrop {
                text: String::from("Drop a file here!"),
                theme: system_theme_mode(),
            },
            Command::none(),
        )
    }

    fn title(&self) -> String {
        String::from("File Drop")
    }

    fn update(&mut self, message: Self::Message) -> Command<Self::Message> {
        match message {
            Message::FileDropped(file) => {
                self.text = format!("Dropped file:\n{}", file.to_string_lossy());
                Command::none()
            }
            Message::SystemThemeMode(theme) => {
                self.theme = theme;
                Command::none()
            }
        }
    }

    fn view(&self) -> Element<Self::Message> {
        Element::new(text(&self.text))
    }

    fn theme(&self) -> Theme {
        self.theme.clone()
    }

    fn subscription(&self) -> Subscription<Self::Message> {
        Subscription::batch([
            subscription::events_with(|event, status| match (event, status) {
                (Event::Window(window::Event::FileDropped(file)), event::Status::Ignored) => {
                    Some(Message::FileDropped(file))
                }
                _ => None,
            }),
            time::every(time::Duration::from_secs(20))
                .map(|_| Message::SystemThemeMode(system_theme_mode())),
        ])
    }
}

fn system_theme_mode() -> Theme {
    match dark_light::detect() {
        dark_light::Mode::Light | dark_light::Mode::Default => Theme::Light,
        dark_light::Mode::Dark => Theme::Dark,
    }
}
```

This requires either one of the following `iced` features to be enabled: `tokio`, `async-std`, or `smol`.

**Comment by mfreeborn** [comment]
Can I confirm that there isn't any easy way to receive notifications directly from the system when the system theme changes? Taking the code from @lucatrv is 99% of the way there, but in order to get the theme to update automatically instantly, we'd need to call dark_light::detect() every few milliseconds, rather than every 60 seconds in the example code. That seems a bit wasteful given that the system theme is not likely to be changed often mid-app.

**Comment by Zocker1999NET** [comment]
Could it be that this issue is already implemented by PR #2343 ?

**Comment by danielstuart14** [comment]
For anyone interested, I've been developing a library that does exactly that: [system-theme](https://github.com/danielstuart14/system-theme)

Right now it supports all the major desktop OSes and is capable of fetching accent color, theme scheme (dark / light), and theme contrast; as well as subscribing to theme changes.
It also has palettes based on the native OS designs (fluent, aqua, adwaita, and breeze) and has integration with iced.

My plan on the next weeks is to implement widget themes (generic for any GUI framework + integration with iced widgets) for all the OS designs already supported. I'm not invested in implementing support for web or mobile, but I'll be glad to review any PRs.

**Comment by lucatrv** [comment]
@danielstuart14 thank you for your contribution, but since PR #2343 has been released my `iced` apps update automatically their theme (dark / light) under Windows following the system settings, what would [`system-theme`](https://github.com/danielstuart14/system-theme) provide in addition to this?

**Comment by danielstuart14** [comment]
> [@danielstuart14](https://github.com/danielstuart14) thank you for your contribution, but since PR [#2343](https://github.com/iced-rs/iced/pull/2343) has been released my `iced` apps update automatically their theme (dark / light) under Windows following the system settings, what would [`system-theme`](https://github.com/danielstuart14/system-theme) provide in addition to this?

system-theme supports async subscription to theme changes, which means your app will follow almost instantly any changes. This also means that you don't need to keep pooling the current theme.
Besides that, it also supports "native" theme palettes that try to follow the UI guidelines to the platform it is running (with high contrast support) and allows you to use the system accent color as configured by the user.

dark-light (the library iced currently uses) is great, but it relies on undocumented APIs on windows (reads registry entries directly) and macos (uses undocumented user defaults entry).

**Comment by pml68** [comment]
Dark-light isn't used anymore, winit's own system theme API and mundy is. That's why changes are now reflected automatically by default, and you can separately subscribe to them as well with https://docs.rs/iced/latest/iced/system/fn.theme_changes.html

**Comment by lucatrv** [comment]
Sorry I understand I did not fully explain... since PR #2343 has been released I'm not using `dark-light` anymore, so with plain `iced` and no theme customization of any sort (I'm not even using `iced::Theme` or define the `theme(&self) -> Theme` function) my `iced` apps sync automatically with system theme settings almost instantly. So there isn't anything special to do now.
IMHO this issue should be closed.

**Comment by WavyEbuilder** [comment]
Does this issue extend to so-called "styles"?

Themes is somewhat complex to define: for example, Qt's "themes" aren't really themes in the traditional sense (they're also referred to as Styles I suppose). This has a vast amount of utility as one can "style"/"theme" an _entire_ suite of Qt applications with incredible granularity as you have control over the paint events for each widget: see https://doc.qt.io/qt-6/qstyle.html.

Is something like that considered out of scope? I can see that as a highly, highly useful feature (and I'm not sure if this would come under the umbrella of theming). It does really dependent on some stable API and ABI and dynamic linking however. Rust at present doesn't have anything of its own (crABI is still a long while off), but there are a few options, for example simply just exporting things using the platform so-called "C ABI" (which although it's a bit of a misnomer will suffice for the purpose of explanation).

The reason for requiring dynamic linking is such that a consumer of a style does _not_ have to be linked against the style at runtime.

Let's say for the sake of argument that I was to install some iced application. This iced application is just its own standalone thing, not tied to any specific desktop or set of interfaces. And let's suppose that I installed it on some "platform", for example the Cosmic Desktop. Without this client application needing to be aware of Cosmic at all, assuming Cosmic was to provide some hypothetical "style library", the application could simply load in this dynamic library and allow it to "register" itself to handle paint events for widgets.

This would allow for visual consistency between a multitude of iced application without them necessarily needing to be aware of each other.

**Comment by upsidedownsweetfood** [comment]
> Does this issue extend to so-called "styles"?
> 
> Themes is somewhat complex to define: for example, Qt's "themes" aren't really themes in the traditional sense (they're also referred to as Styles I suppose). This has a vast amount of utility as one can "style"/"theme" an _entire_ suite of Qt applications with incredible granularity as you have control over the paint events for each widget: see https://doc.qt.io/qt-6/qstyle.html.
> 
> Is something like that considered out of scope? I can see that as a highly, highly useful feature (and I'm not sure if this would come under the umbrella of theming). It does really dependent on some stable API and ABI and dynamic linking however. Rust at present doesn't have anything of its own (crABI is still a long while off), but there are a few options, for example simply just exporting things using the platform so-called "C ABI" (which although it's a bit of a misnomer will suffice for the purpose of explanation).
> 
> The reason for requiring dynamic linking is such that a consumer of a style does _not_ have to be linked against the style at runtime.
> 
> Let's say for the sake of argument that I was to install some iced application. This iced application is just its own standalone thing, not tied to any specific desktop or set of interfaces. And let's suppose that I installed it on some "platform", for example the Cosmic Desktop. Without this client application needing to be aware of Cosmic at all, assuming Cosmic was to provide some hypothetical "style library", the application could simply load in this dynamic library and allow it to "register" itself to handle paint events for widgets.
> 
> This would allow for visual consistency between a multitude of iced application without them necessarily needing to be aware of each other.

From what I've understood, the cosmic theming engine right now is built in as a crate, so at build time, and loads a theme TOML file (?)

I honestly would hope that maybe Iced would break off from the modern static CSS theming, or whatever, and have a more dynamic QT-styled way of handling styles.

I was currently in the process of prototyping a KDE union crate for Iced, and this brought me to this issue, I figure that a C ABI interface would be the best compromise to glue at runtime any styling engine one could wish for, which, as far as I know, was how GTK2 worked too.

I guess that someone could implement this as an iced theme, it doesn't really have to be included in the base library, is someone already working on this, just to avoid duplicating effort

**Comment by WavyEbuilder** [comment]
> > Does this issue extend to so-called "styles"?
> > Themes is somewhat complex to define: for example, Qt's "themes" aren't really themes in the traditional sense (they're also referred to as Styles I suppose). This has a vast amount of utility as one can "style"/"theme" an _entire_ suite of Qt applications with incredible granularity as you have control over the paint events for each widget: see https://doc.qt.io/qt-6/qstyle.html.
> > Is something like that considered out of scope? I can see that as a highly, highly useful feature (and I'm not sure if this would come under the umbrella of theming). It does really dependent on some stable API and ABI and dynamic linking however. Rust at present doesn't have anything of its own (crABI is still a long while off), but there are a few options, for example simply just exporting things using the platform so-called "C ABI" (which although it's a bit of a misnomer will suffice for the purpose of explanation).
> > The reason for requiring dynamic linking is such that a consumer of a style does _not_ have to be linked against the style at runtime.
> > Let's say for the sake of argument that I was to install some iced application. This iced application is just its own standalone thing, not tied to any specific desktop or set of interfaces. And let's suppose that I installed it on some "platform", for example the Cosmic Desktop. Without this client application needing to be aware of Cosmic at all, assuming Cosmic was to provide some hypothetical "style library", the application could simply load in this dynamic library and allow it to "register" itself to handle paint events for widgets.
> > This would allow for visual consistency between a multitude of iced application without them necessarily needing to be aware of each other.
> 
> From what I've understood, the cosmic theming engine right now is built in as a crate, so at build time, and loads a theme TOML file (?)

Hmm, so I'm guessing then we're limited to what the TOML file is capable of serializing, which is unfortunately less than what even some form of embedded dynamic scripting language exposed for paint events (which whilst not ideal in my opinion would still actually be a fairly reasonable compromise).

> I honestly would hope that maybe Iced would break off from the modern static CSS theming, or whatever, and have a more dynamic QT-styled way of handling styles.
> 
> I was currently in the process of prototyping a KDE union crate for Iced, and this brought me to this issue, I figure that a C ABI interface would be the best compromise to glue at runtime any styling engine one could wish for, which, as far as I know, was how GTK2 worked too.

The best of luck with union and iced, that sounds really fantastic! However, I'm not sure as to how using the platform ABI would help here without changes in _iced_ itself as...

> I guess that someone could implement this as an iced theme, it doesn't really have to be included in the base library, is someone already working on this, just to avoid duplicating effort

From what I can see here and above we're still limited the toml format at present sadly :/.

So whilst that's probably enough to handle colours and some other basic/high-level "theme" options, it doesn't really seem to handle what I would classify as a _style_ given the limitations of serializing in such a format. I think we'd have to expose over the platform ABI some way of handling paint event for themes such that DSOs may then handle the paint event for each widget.

It's a fair amount of work, but before that can even be considered, one would need to know:
1. Is this within scope and/or something iced is open to/interested in?
2. If so, what is the scope of the aforementioned styles and how far do they extend.

**Comment by upsidedownsweetfood** [comment]
> > > Does this issue extend to so-called "styles"?
> > > Themes is somewhat complex to define: for example, Qt's "themes" aren't really themes in the traditional sense (they're also referred to as Styles I suppose). This has a vast amount of utility as one can "style"/"theme" an _entire_ suite of Qt applications with incredible granularity as you have control over the paint events for each widget: see https://doc.qt.io/qt-6/qstyle.html.
> > > Is something like that considered out of scope? I can see that as a highly, highly useful feature (and I'm not sure if this would come under the umbrella of theming). It does really dependent on some stable API and ABI and dynamic linking however. Rust at present doesn't have anything of its own (crABI is still a long while off), but there are a few options, for example simply just exporting things using the platform so-called "C ABI" (which although it's a bit of a misnomer will suffice for the purpose of explanation).
> > > The reason for requiring dynamic linking is such that a consumer of a style does _not_ have to be linked against the style at runtime.
> > > Let's say for the sake of argument that I was to install some iced application. This iced application is just its own standalone thing, not tied to any specific desktop or set of interfaces. And let's suppose that I installed it on some "platform", for example the Cosmic Desktop. Without this client application needing to be aware of Cosmic at all, assuming Cosmic was to provide some hypothetical "style library", the application could simply load in this dynamic library and allow it to "register" itself to handle paint events for widgets.
> > > This would allow for visual consistency between a multitude of iced application without them necessarily needing to be aware of each other.
> > 
> > 
> > From what I've understood, the cosmic theming engine right now is built in as a crate, so at build time, and loads a theme TOML file (?)
> 
> Hmm, so I'm guessing then we're limited to what the TOML file is capable of serializing, which is unfortunately less than what even some form of embedded dynamic scripting language exposed for paint events (which whilst not ideal in my opinion would still actually be a fairly reasonable compromise).
> 
> > I honestly would hope that maybe Iced would break off from the modern static CSS theming, or whatever, and have a more dynamic QT-styled way of handling styles.
> > I was currently in the process of prototyping a KDE union crate for Iced, and this brought me to this issue, I figure that a C ABI interface would be the best compromise to glue at runtime any styling engine one could wish for, which, as far as I know, was how GTK2 worked too.
> 
> The best of luck with union and iced, that sounds really fantastic! However, I'm not sure as to how using the platform ABI would help here without changes in _iced_ itself as...
> 
> > I guess that someone could implement this as an iced theme, it doesn't really have to be included in the base library, is someone already working on this, just to avoid duplicating effort
> 
> From what I can see here and above we're still limited the toml format at present sadly :/.
> 
> So whilst that's probably enough to handle colours and some other basic/high-level "theme" options, it doesn't really seem to handle what I would classify as a _style_ given the limitations of serializing in such a format. I think we'd have to expose over the platform ABI some way of handling paint event for themes such that DSOs may then handle the paint event for each widget.
> 
> It's a fair amount of work, but before that can even be considered, one would need to know:
> 
>     1. Is this within scope and/or something iced is open to/interested in?
> 
>     2. If so, what is the scope of the aforementioned styles and how far do they extend.

preface: Sorry for my poor english.


Supposedly, we wouldn't be stuck with the toml theming "engine" that the cosmic team made, since, if we were to adopt the C-ABI as a standard for the  said engines, any could be plugged in at runtime, and, ideally, do whatever they want, as long as they output something that matches to an iced Theme.

The engine itself could be swapped or chosen either with DBUS or an environment variable, and that would allow pretty much feature parity, at least, supposedly, with what most theming engines already do.

The only issue I have with it, is that we'd need C glue to this, and the nice part of it is that we wouldn't need to edit the code of iced at all for a simple implementations, application developers could opt-in and just install a crate, and then wire some code up in a way that would allow their application to use any given engine.

 we already have
```rust
fn theme(&self, state: &State) -> Option<Theme>;
```
one would just need to provide a theme function that sets the basics up.

Still, I think this will require more thought, but Hecrj may be more willing to include something like this in the iced library once he sees enough community commitment? I'm not totally sure, but could be a decent way to get this in.

**Comment by upsidedownsweetfood** [comment]
I'm starting to work on this, progress will be slow probably as this will be developed in my free time, if anyone is interested, I would definitely like a hand; my matrix is `@creamcake:syn.absentmind.it`, although, can communicate via any place you want.

Right now I'm writing a prototype for this, it's a rust crate that loads a given dynamic library, for now via an environment variable, that communicates via the C-ABI ( I'm not sure if I want it to use the unstable rust one ), that exports a load theme function, that outputs a string that will be serialized by the rust side an then converted to an iced theme definition.

I really want to support dynamic reloading of the theme, but I'm not totally sure for how to handle that ( maybe I could just produce an iced message, and the developer will choose if they want to handle it. )

EDIT: I think that my approach is decent, even if a bit over-complicated as It would be nice to let the user choose any theming engine they want, let's say, for example, that I want a gtk theme as my iced ones, I could write the engine myself as a dynamic library, load a gtk theme via CSS, and my iced application would't know anything about it.
And it's also not as static as the cosmic's implementation.

Further edit: Of course this is not really comparable to what QT does, styles there are way more advanced, but this should be a good middle ground between static themes and highly dynamic ones.
